### DISCUSSION TITLE
Increasing the camera range with double float precision

### DISCUSSION THREAD

**Discussion by AlanTheSane** [ivoyager/forum#15] [General]
Found this project while looking into achieving solar system scale in Godot, this will be a very useful resource! I noticed objects/orbits disappearing when zoomed in close to Juno, and found the comment in code noting the camera limitation (max far multiplier of 1e6). 

For those interested this can be increased if you [build Godot](https://docs.godotengine.org/en/stable/contributing/development/compiling/introduction_to_the_buildsystem.html) with [double float precision](https://docs.godotengine.org/en/4.2/classes/class_float.html#:~:text=it%20can%20be%20changed%20to%20use%2064%2Dbit%20doubles%20if%20Godot%20is%20compiled%20with%20the%20precision%3Ddouble%20option.) enabled. Annoyingly it isn't enabled by default or available pre-built and my crappy laptop took well over an hour to build Godot from source, but once done I was able to increase the camera FAR_MULTIPLIER to 1e14 before it started breaking again.

This will presumably help avoid accuracy issues in other areas of the code but I'm surprised how much you can manage with just 32 bit floats.

**Comment by AlanTheSane** [comment]
Turning this into a bit of a feature request...

It might be an idea to use [OS.has_feature("double")](https://docs.godotengine.org/en/latest/tutorials/export/feature_tags.html) to conditionally set this and other values that would benefit in double precision builds.

**Comment by charliewhitfield** [comment]
Thanks for trying this. I was pessimistic that double-float build would improve the near/far range. I figured that limit was on the GPU side. I don't really understand what the double-float build changes: I've done experiments that indicated to me that GDScript math uses double precision if you avoid Vector built-ins.

I'll definitely try this going forward. I expect to be updating for 4.3 in September. I'll make the Godot builds available for download with our releases downloads.

**Comment by AlanTheSane** [comment]
I actually misunderstood at first and thought the option to build for double precision meant all floats were single precision by default. You are correct, and it even states it clearly on one of the pages I linked, a "float" is always 64-bit while stuff like vectors use 32-bit floats internally for efficiency. 

My assumption is that even if you are directly feeding a float to a node like the camera, they either follow the same efficiency principle or they themselves use vectors internally, truncating everything they touch long before the GPU gets involved.

Following on from [your comment](https://github.com/ivoyager/ivoyager_core/issues/1#issuecomment-2243423555) about reasons the ring shader might be breaking... The fact that float values were always 64-bit indicates to me they _probably_(?) aren't the problem. `shader_frame_data` gets passed in as a vector4 so I tried splitting that up into floats and feeding each value separately but it made no difference. I have come to realise that shaders are a lot harder to debug than other code :( .

**Comment by AlanTheSane** [comment]
Rings are broken regardless of double/single build, its a 4.3 thing. My bad. 

When I did the custom build I just used the main branch at the time... so somewhere between beta3 and RC1. I seemingly got very lucky as beta3 has a lot of issues just loading the project for me, while RC1 mostly works (except rings). Add to this, they aren't _completely_ missing as I first thought. You can see a bit of the rings at close in viewing angles.

**Comment by charliewhitfield** [comment]
Godot 4.3 breaks our "1d" ring texture mipmaps. I guess it was a bit iffy that this worked before. We have 14495 x 2 textures that were presumably reduced as follows:

14495 x 2 (LOD level = 0)
7247 x 1 (LOD level = 1)
3623 x 1 (2)
1811 x 1 (3)
905 x 1 (4)
etc... (the shader used up to LOD = 8, always sampling at y=0)

(Perhaps I should have know this was iffy in 4.2.x because it required the 2 pixel width in the original.)

HOWEVER, in Godot 4.3 only LOD 0 and 1 are available. Apparently, it stops reducing one dimension when the other dimension is fully reduced to 1 pixel. @AlanTheSane, do you happen to know if this is "industry standard"? Or should I report this as an issue?

The quickest solution would be to stretch the images from 14495 x 2 to 14495 x 256. However, this increases image sizes from ~20kb to >2M. I think what I'll need to do is generate the LOD levels as separate png images.

**Comment by AlanTheSane** [comment]
I have no idea of what's standard, or if this was a deliberate change. That's a neat way of getting a high resolution ring with a small texture so hopefully it's not too difficult to remedy.

I have not wrapped my head around shaders yet. I was following a (v3.5) example a couple weeks ago for something that sampled a 1D gradient, but since 4.x sampling the texture in shader at x=0.0 returns the colour at x=1.0... Don't know why this would be the intended behaviour but couldn't find anyone complaining about it being broken. Sampling at x=0.01 as a workaround felt too janky to be intended.

**Reply by charliewhitfield** [reply]
Rings are fixed now in the _master_ branch for Godot 4.3 (the editor will ask to update assets). Now on to shadows...

> , but since 4.x sampling the texture in shader at x=0.0 returns the colour at x=1.0... Don't know why this would be the intended behaviour but couldn't find anyone complaining about it being broken. Sampling at x=0.01 as a workaround felt too janky to be intended.

Something in the fragment shader (possibly interpolation related) may bump your 0.0 or 1.0 to just a little above or below the integral value. Also, FRAGCOORD.xy has a +0.5 offset from the pixel integer coordinate: so here you might think it's 0.0 but it's really 0.5. Given this, it's possible that your use of int() or round() somewhere is not doing what you think it should.

**Reply by AlanTheSane** [reply]
I figured it was something to do with sampling at the edge rather than a specific pixel. Going back to it now I've figured it out. It seems that either the pixel offset changed or texture repeating is enabled by default now when it didn't used to be(?).

I found people talking about enabling texture repeating with a `repeat_enable` hint on the uniform, so I tried the inverse with `repeat_disable` and now sampling at 0 doesn't return a mix between opposite ends of the gradient.
